src/Features/LanguageServer/Protocol/Handler/MapCode/MapCodeHandler.cs

```diff
@@ -0,0 +1,152 @@
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler;
+
+[ExportCSharpVisualBasicStatelessLspService(typeof(MapCodeHandler)), Shared]
+[Method(WorkspaceMapCodeName)]
+internal sealed partial class MapCodeHandler : ILspServiceRequestHandler<MapCodeParams, LSP.WorkspaceEdit?>
+{
+    public const string WorkspaceMapCodeName = "workspace/mapCode";
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public MapCodeHandler()
+    {
+    }
+
+    public bool MutatesSolutionState => false;
+    public bool RequiresLSPSolution => true;
+
+    public async Task<WorkspaceEdit?> HandleRequestAsync(MapCodeParams request, RequestContext context, CancellationToken cancellationToken)
+    {
+        Contract.ThrowIfNull(context.Solution);
+
+        //TODO: handle request.Updates if not empty
+        if (request.Updates is not null)
+        {
+            throw new NotImplementedException("mapCode Request failed: additional workspace 'Update' is currently not supported");
+        }
+
+        using var _ = PooledDictionary<Uri, TextEdit[]>.GetInstance(out var uriToEditsMap);
+        foreach (var codeMapping in request.Mappings)
+        {
+            var mappingResult = await MapCodeAsync(codeMapping).ConfigureAwait(false);
+
+            if (mappingResult is not (Uri uri, TextEdit[] textEdits))
+            {
+                // Failed the entire request if any of the sub-requests failed
+                context.TraceWarning("mapCode Request failed: a sub-request failed");
+                return null;
+            }
+
+            if (!uriToEditsMap.TryAdd(uri, textEdits))
+                throw new ArgumentException($"mapCode sub-request for {uri} failed: multiple MapCodeMappings for the same document is not supported.");
+        }
+
+        // return a combined WorkspaceEdit
+        if (context.GetRequiredClientCapabilities().Workspace?.WorkspaceEdit?.DocumentChanges is true)
+        {
+            return new WorkspaceEdit
+            {
+                DocumentChanges = uriToEditsMap.Select(kvp => new TextDocumentEdit
+                {
+                    TextDocument = new OptionalVersionedTextDocumentIdentifier { Uri = kvp.Key },
+                    Edits = kvp.Value,
+                }).ToArray()
+            };
+        }
+        else
+        {
+            return new WorkspaceEdit
+            {
+                Changes = uriToEditsMap.ToDictionary(kvp => ProtocolConversions.GetDocumentFilePathFromUri(kvp.Key), kvp => kvp.Value)
+            };
+        }
+
+        async Task<(Uri, TextEdit[])?> MapCodeAsync(LSP.MapCodeMapping codeMapping)
+        {
+            var textDocument = codeMapping.TextDocument
+                ?? throw new ArgumentException($"mapCode sub-request failed: MapCodeMapping.TextDocument not expected to be null.");
+
+            if (context.Solution.GetDocument(textDocument) is not Document document)
```

Comment:
this request is not associated with one document (i.e. the handler isILspServiceRequestHandler instead of ILspServiceDocumentRequestHandler). It can contain multiple sub-reqeusts each targeting a document. I was not involved in the design of the protocol so can't really comment on the reason for this FYI @allisonchou